src/include/grpc_transcoding/message_reader.h

```diff
@@ -78,6 +82,15 @@ class MessageReader {
   // false.
   bool Finished() const { return finished_ || !status_.ok(); }
 
+  // Returns the last gRPC message delimiter processed by this reader.
+  // The caller should only read `kGrpcDelimiterByteSize` bytes from this
+  // pointer.
+  //
+  // The caller should only call this method if `NextMessage()` provides a
+  // non-nullptr result and `Status()` is OK. Otherwise the underlying memory
+  // contents may be uninitialized.
+  unsigned char* LastDelimiter() { return delimiter_; }
```

Comment:
This usage is not very clear.   How about adding a new NextMessage() function as

std::unique_ptr<::google::protobuf::io::ZeroCopyInputStream> NextMessageWithDelimitor(unsigned char*delimitor_buf);

// only a full message is read, both msg and delimitor is valid.

the old function drop the delimitor and the new function preserve it.  Will that serve your purpose?

Reply:
Yes, i like this idea.

I am having trouble with the unit tests. Can you double check if there are any lifetime issues in my implementation? For some reason the stack-allocated buffer is failing ~half of the tests.

The previous implementation that kept the buffer owned in the MessageReader worked fine.

Reply:
```
test/message_reader_test.cc:116
Expected equality of these values:
  delimiter_string
    Which is: "\0\0\0\0\0"
  next_expected_->delimiter
    Which is: "\0\0\0\0\r"

test/test_common.cc:169
Failed
  |  SimpleMessage"

test/test_common.cc:169
Failed
Failed for the following partition "\0\0\0\0
  |  SimpleMessage"

test/message_reader_test.cc:234
Value of: tc->Test(2, 1.0)
  Actual: false
Expected: true
```

Reply:
replied inline in the code

Reply:
Thanks, makes a lot of sense. Changed the API so class maintains ownership of the underlying buffer.